substrate/frame/staking-async/ah-client/src/lib.rs

```diff
@@ -632,29 +692,48 @@ pub mod pallet {
 					);
 				}
 
-				// prepare an `Offence` instance for the XCM message. Note that we drop the
-				// identification.
 				let (offender, _full_identification) = offence.offender;
 				let reporters = offence.reporters;
-				offenders_and_slashes.push(rc_client::Offence {
-					offender,
-					reporters,
-					slash_fraction: *fraction,
-				});
+
+				match mode {
+					OperatingMode::Buffered => {
+						// In `Buffered` mode, we buffer the offences for later processing.
+						// We only keep the highest slash fraction for each offender.
+						BufferedOffences::<T>::mutate(|buffered| {
+							let session_offences = buffered.entry(slash_session).or_default();
+							let entry = session_offences.entry(offender);
+
+							entry
+								.and_modify(|existing| {
+									if existing.slash_fraction < *fraction {
+										*existing = BufferedOffence {
+											reporter: reporters.first().cloned(),
+											slash_fraction: *fraction,
+										};
+									}
+								})
+								.or_insert(BufferedOffence {
+									reporter: reporters.first().cloned(),
+									slash_fraction: *fraction,
+								});
+						});
+					},
+					OperatingMode::Active => {
+						// prepare an `Offence` instance for the XCM message. Note that we drop the
+						// identification.
+						offenders_and_slashes_message.push(rc_client::Offence {
+							offender,
+							reporters,
+							slash_fraction: *fraction,
+						});
+					},
+					_ => {},
```

Comment:
I freaked out here for a sec, thinking that we forgot to pass the offence forward to `Fallback` in the last arm -- We can mark this arm as `unreachable!()` or have a comment that it is already handled before. Or reorganize the code into one big `match`.